rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,226 @@
+the same cached tensor, and we had a promise-future mechanism to allow only one
+thread to initiate the computation of constant tensor, and other threads to wait
+until the computation completes.
+
+![DNNL backend existing constant tensor cache](images/dnnl-backend-constant-cache.png)
+
+We has exposed two APIs to enable/disable the constant cache and
+return the enabling status:
+
+```cpp
+void set_constant_tensor_cache(int flag);
+int get_constant_tensor_cache();
+```
+
+In some memory limited scenarios, users may still want to cache a part of
+constant tensors to improve performance, but don't want the cached capacity to
+grow infinitely and consume too much memory. In this document, we propose to
+introduce a constant tensor cache capacity control API to address above
+requests.
+
+## Proposal
+
+### Dedicated cache for different engine kind
+In oneDNN Graph programming model, different graph can have different engine
+kind and simultaneously run on different engine, like cpu or gpu. Different kind
+of engine usually has different memory limitation. 
+
+So, we propose to use dedicated cache for different engine kind, and let users
+set different capacity for different engine kind separately.
+
+For multiple devices scenarios, like multi-GPU, whether to use dedicated cache
+for each device is still an open question, we will discuss it in later sections.
+
+### Unified cache for all backend
+We propose to implement an unified constant tensor cache utils in the oneDNN
+Graph interface layer. Each backend should use this interface to cache all their
+constant tensors in to the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+The key and value design of unified constant tensor cache utils:
+
+- Cache key: The unified cache key can be composed of two parts: backend specific
+  key and backend id. The backend id is unique among backends. The backend
+  specific key can be defined in each backends, to achieve more flexibility.
+- Cache value: The unified cache value can be a base buffer class like following,
+  and each backend can inherit this base class and subclass it.
+
+```cpp
+class constant_buffer_t{
+public:
+    virtual size_t get_size() const;
+    virtual ~constant_buffer_t();
+
+private:
+    void *data_;
```

Comment:
How to access this data pointer since it's private?

Reply:
will change to protected and add getter method.

Reply:
improved